### PR TITLE
Set up CI with Azure Pipelines

### PR DESCRIPTION
**Pull request by StephenHodgson**

### CODE REVIEW

File: stages/build.yml [hunk 1]
```diff
@@ -0,0 +1,12 @@
+stages:
+- stage: Build for ${{ format('{0} ({1})', parameters.platform, parameters.image) }}
+  pool:
+    vmImage: ${{ parameters.image }}
+  variables:
+    unityExtPath: ${{  }}
```
Comment (Jarodshow) [inline stages/build.yml:6]: Does anyone know the path on linux?
Windows is `\Editor\Unity.exe`
MacOS is `/Unity.app/Contents/MacOS/Unity`

Comment (StephenHodgson) [reply stages/build.yml:6]: You should be able to get this from the Unity Powershell module we install.

Comment (StephenHodgson) [reply stages/build.yml:6]: This shouldn't have to be a variable we set

File: jobs/release-unity.yml [hunk 2]
```diff
@@ -0,0 +1,11 @@
+jobs:
+- job: Release Unity License
+  steps:
+  - powershell: |
+      'if ( Test-Path $(System.DefaultWorkingDirectory)\$(project.name)\Builds\Logs\activation.log -Type Leaf ) {
+      Write-Host "Returning License..."
+      $editorInstance = Get-UnitySetupInstance | Select-UnitySetupInstance -Latest
+      $editorPath = $editorInstance.Path + "\Editor\Unity.exe"
```
Comment (StephenHodgson) [inline jobs/release-unity.yml:8]: Here's where we get the editor instance path.

File: azure-pipelines.yml [hunk 3]
```diff
@@ -0,0 +1,8 @@
+variables:
+- name: project.folder
+  value: XRTK
+- name: project.name
+  value: XRTK-Core
```
Comment (StephenHodgson) [inline azure-pipelines.yml:5]: Is this just a duplicate of what's in the `azure-pipelines-core.yml`?

Can we remove it?

File: azure-pipelines-core.yml [hunk 4]
```diff
@@ -0,0 +1,41 @@
+resources:
+  repositories:
+  - repository: templates
+    type: github
+    name: XRTK/Azure-Templates
```
Comment (StephenHodgson) [inline azure-pipelines-core.yml:5]: I think the name needs to be `XRTK/Azure-Pipelines`

File: jobs/install-unity.yml [hunk 5]
```diff
@@ -0,0 +1,45 @@
+jobs:
+- job: Install Unity
+  steps:
+  - checkout: self
+    submodules: $(project.isCore)
```
Comment (StephenHodgson) [inline jobs/install-unity.yml:5]: we should always pull submodules

File: jobs/install-unity.yml [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+          Write-Error "Failed to find unity credentials"
+          exit 1
+      }
+      
+      $username = $unityCredentials.username
+      $password = $unityCredentials.password
+      $serial = $unityCredentials.serial
+      
+      $editorInstance = Get-UnitySetupInstance | Select-UnitySetupInstance -Latest
+      $editorPath = $editorInstance.Path + "\Editor\Unity.exe"
```
Comment (StephenHodgson) [inline jobs/install-unity.yml:33]: Here's where we get the editor path the first time. Let's set a variable here in pipelines so we can reuse it later.

File: pipieline-build.yml [hunk 7]
```diff
@@ -0,0 +1,34 @@
+parameters:
+- name: os
+  type: object
+  default:
+    windows:
+      image: vs2017-win2016
+      components: '"Windows", "Windows_IL2CPP"'
```
Comment (StephenHodgson) [inline pipieline-build.yml:7]: I'd like to define these components in each package template

File: stages/sync.yml [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+  default: vs2017-win2016
+
+stages:
+- stage: Sync
+  condition: ${{ parameters.shouldRun }}
+
+  pool:
+    vmImage: ${{ parameters.vmImage }}
+  jobs:
+  - job: DocFX Build
```
Comment (StephenHodgson) [inline stages/sync.yml:16]: We should only do the docfx build in core

File: azure-pipelines-core.yml [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+  value: true
+- template: global-group.yml@templates
+- template: publish-group.yml@templates
+
+
+trigger:
+  branches:
+    include:
+    - master
+    - upm
```
Comment (StephenHodgson) [inline azure-pipelines-core.yml:24]: I don't think we want a upm trigger

File: pipieline-build.yml [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+- ${{ each pair in parameters.os }}:
+  - template: stages/build.yml
+    parameters:
+      platform: ${{ pair.key }}
+      ${{ insert }}: ${{ pair.value }}
+      unityJobs: ${{ parameters.unityJobs }}
+    
+- template: stages/sync.yml
+  parameters:
+    shouldRun: ${{ if eq(variables['build.sourceBranchName'], 'upm') }}
```
Comment (Jarodshow) [inline pipieline-build.yml:33]:
```suggestion
    shouldRun: ${{ if eq(variables['build.reason'], 'PullRequest') }}
```

File: stages/sync.yml (hunk 8 above)
Comment (Jarodshow) [reply stages/sync.yml:16]:
```suggestion
  - job: DocFX Build
    condition: $(project.isCore)
```

File: pipieline-build.yml (hunk 7 above)
Comment (Jarodshow) [reply pipieline-build.yml:7]: These can be overridden per pipeline, is that what you mean?

File: azure-pipelines-core.yml (hunk 4 above)
Comment (Jarodshow) [inline azure-pipelines-core.yml:5]:
```suggestion
    name: XRTK/Azure-Pipelines
```

File: azure-pipelines-core.yml (hunk 9 above)
Comment (Jarodshow) [inline azure-pipelines-core.yml:24]:
```suggestion
```

File: jobs/install-unity.yml (hunk 5 above)
Comment (Jarodshow) [reply jobs/install-unity.yml:5]: Even for oculus, wmr, etc.?

Comment (StephenHodgson) [reply jobs/install-unity.yml:5]: yes

File: azure-pipelines-core.yml [hunk 11]
```diff
@@ -0,0 +1,42 @@
+name: XRTK-$(System.PullRequest.PullRequestNumber)$(Rev:.r)
```
Comment (StephenHodgson) [inline azure-pipelines-core.yml:1]: I'd like to be able to fallback to a date if no Pull Request number is available. Is that possible?

File: stages/build.yml (hunk 1 above)
Comment (Jarodshow) [reply stages/build.yml:6]: `Start-UnityEditor` seems so be the solution here...

Comment (StephenHodgson) [reply stages/build.yml:6]:
```
      $editorInstance = Get-UnitySetupInstance | Select-UnitySetupInstance -Latest
      $editorPath = $editorInstance.Path + "\Editor\Unity.exe"
```

File: azure-pipelines-core.yml (hunk 11 above)
Comment (Jarodshow) [reply azure-pipelines-core.yml:1]: absolutely
```suggestion
- ${{ if eq(variables[build.reason], 'IndividualCI') }}:
  name: XRTK-$(Date:yyyyMMdd)$(Rev:.r)
- ${{ if nq(variables[build.reason], 'IndividualCI') }}:
  name: XRTK-$(System.PullRequest.PullRequestNumber)$(Rev:.r)
```

is this format good?

Comment (StephenHodgson) [reply azure-pipelines-core.yml:1]: yup I think that works for me

File: jobs/build-unity.yml [hunk 12]
```diff
@@ -0,0 +1,35 @@
+parameters:
+- name: runTests
+  type: boolean
+  default: true
```
Comment (StephenHodgson) [inline jobs/build-unity.yml:4]: I don't think this should be optional

File: stages/build.yml [hunk 13]
```diff
@@ -0,0 +1,10 @@
+stages:
+- stage: Build for ${{ format('{0} ({1})', parameters.platform, parameters.image) }}
+  pool:
+    vmImage: ${{ parameters.image }}
+  jobs:
+  - template: jobs/install-unity.yml
+  - template: jobs/build-unity.yml
```
Comment (StephenHodgson) [inline stages/build.yml:7]: Is there a way to enforce these to sun sequentially and not at the same time?

File: stages/sync.yml [hunk 14]
```diff
@@ -0,0 +1,91 @@
+parameters:
+- name: shouldRun
+  type: boolean
+  default: true
+- name: vmImage
+  type: string
+  default: vs2017-win2016
+
+stages:
+- stage: Sync
```
Comment (StephenHodgson) [inline stages/sync.yml:10]: Can we rename this file  and stage to `publish.yml`?

File: pipelines/azure-pipelines-wmr.yml [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+  repositories:
+  - repository: templates
+    type: github
+    name: XRTK/Azure-Pipelines
+
+variables:
+- name: project.folder
+  value: XRTK
+- name: project.name
+  value: XRTK-Lumin
```
Comment (StephenHodgson) [inline pipelines/azure-pipelines-wmr.yml:11]: Let's hold off on creating pipelines for all the platforms yet. I wanna get just core working first

File: stages/build.yml (hunk 13 above)
Comment (Jarodshow) [reply stages/build.yml:7]:
```
  strategy:
    parallel: 1
```
has the same effect I believe when added to all the jobs

File: pipieline-build.yml [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+  type: jobList
+  default: []
+
+pool:
+  name: 'Azure Pipelines'
+  demands: 
+    - unity -equals true
+
+stages:
+- ${{ each pair in parameters.os }}:
```
Comment (StephenHodgson) [inline pipieline-build.yml:29]: [Wasn't able to get this to work as expected](https://dev.azure.com/xrtk/Mixed%20Reality%20Toolkit/_build/results?buildId=3544&view=results)

Comment (Jarodshow) [reply pipieline-build.yml:29]: [this works now :)](https://i.imgur.com/7aFELxA.png)

Comment (StephenHodgson) [reply pipieline-build.yml:29]: I wonder why this doesn't work for me.

Are you passing any parameters into this template?

**Pull request closed without merging**